suite/tests/linux/syscall_pwait.c

```diff
@@ -93,119 +126,141 @@ main(int argc, char *argv[])
 
     INIT();
 
-    sleeptime.tv_sec = 0;
-    sleeptime.tv_nsec = 500 * 1000 * 1000;
+    sleeptime.tv_sec = 1;
+    sleeptime.tv_nsec = 0;
 
     intercept_signal(SIGUSR1, (handler_3_t)signal_handler, true);
     intercept_signal(SIGUSR2, (handler_3_t)signal_handler, true);
     print("handlers for signals: %d, %d\n", SIGUSR1, SIGUSR2);
     sigemptyset(&new_set);
+    sigaddset(&new_set, SIGUSR2);
     sigaddset(&new_set, SIGUSR1);
-    sigprocmask(SIG_BLOCK, &new_set, NULL);
+    pthread_sigmask(SIG_BLOCK, &new_set, NULL);
+    print("signal blocked: %d\n", SIGUSR2);
     print("signal blocked: %d\n", SIGUSR1);
 
     sigset_t test_set;
     sigfillset(&test_set);
     sigdelset(&test_set, SIGUSR1);
     sigdelset(&test_set, SIGUSR2);
 
-    if (kick_off_child_signals())
-        return 0;
-
     int epoll_fd = epoll_create1(EPOLL_CLOEXEC);
     struct epoll_event events;
 
+    ready_to_listen = create_cond_var();
+    pthread_t main_thread = pthread_self();
+
     print("Testing epoll_pwait\n");
 
-    int count = 0;
-    while (count++ < 2) {
-        /* XXX i#3240: DR currently does not handle the atomicity aspect of this system
-         * call. Once it does, please include this in this test or add a new test.
-         */
-        sigset_t pre_syscall_set = {
-            0, /* Set padding to 0 so we can use memcmp */
-        };
-        sigprocmask(SIG_SETMASK, NULL, &pre_syscall_set);
-        if (epoll_pwait(epoll_fd, &events, 24, -1, &test_set) == -1) {
-            if (errno != EINTR)
-                perror("expected EINTR");
-        } else {
-            perror("expected interruption of syscall");
-        }
-        sigset_t post_syscall_set = {
-            0, /* Set padding to 0 so we can use memcmp */
-        };
-        sigprocmask(SIG_SETMASK, NULL, &post_syscall_set);
-        if (memcmp(&pre_syscall_set, &post_syscall_set, sizeof(pre_syscall_set)) != 0) {
-            print("sigmask mismatch");
-            exit(1);
+    int count;
+    for (int i = 0; i < 2; ++i) {
+        count = 0;
+        while (count++ < 2) {
+            /* XXX i#3240: DR currently does not handle the atomicity aspect of this
+             * system call. Once it does, please include this in this test or add a new
+             * test.
+             */
+            sigset_t pre_syscall_set = {
+                0, /* Set padding to 0 so we can use memcmp */
+            };
+            /* delay == false sends the signal before the system call is executed such
+             * that the signal is in pending state once we start the call.
+             * delay == true adds a delay before sending the signal such that the signal
+             * arrives while we are in the system call, but there is no check to verifiy
+             * whether it arrived "late enough".
+             */
+            bool delay = i == 0 ? false : true;
+            kick_off_child_signal(count, main_thread, !delay);
```

Comment:
It is confusing to negate the bool here when you completely control the construction of it: could we rename it to "immediate" or sthg and avoid the negation.

Reply:
Done.